File: src/main/connection.cc
/** Copyright 2020 Alibaba Group Holding Limited.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "neug/main/connection.h"

#include <exception>

#include "neug/main/execution_slot.h"
#include "neug/utils/access_mode.h"
#include "neug/utils/exception/exception.h"
#include "neug/utils/load_profiler.h"
#include "neug/utils/yaml_utils.h"

namespace neug {

Connection::Connection(std::unique_ptr<ExecutionSlot> execution_slot,
                       CloseCallback on_close)
    : execution_slot_(std::move(execution_slot)),
      on_close_(std::move(on_close)) {
  CHECK(execution_slot_ != nullptr);
}

Connection::~Connection() { Close(); }

std::string Connection::GetSchema() const {
  if (IsClosed()) {
    LOG(ERROR) << "Connection is closed, cannot get schema.";
    THROW_RUNTIME_ERROR("Connection is closed, cannot get schema.");
  }
  if (transaction_context_.IsRollbackOnly()) {
    if (transaction_context_.IsBulkLoad()) {
      THROW_TX_STATE_CONFLICT(
          "Bulk-load session failed; RollbackBulkLoad() is required before "
          "GetSchema.");
    }
    THROW_TX_STATE_CONFLICT(
        "Transaction is rollback-only; Rollback() is required before "
        "GetSchema.");
  }
  if (transaction_context_.IsActive()) {
    auto yaml = transaction_context_.schema().to_yaml();
    return get_json_string_from_yaml(yaml.value()).value();
  }
  return execution_slot_->GetSchema();
}

void Connection::Close() {
  if (is_closed_.load(std::memory_order_relaxed)) {
    LOG(WARNING) << "Connection is already closed.";
    return;
  }
  LOG(INFO) << "Closing connection.";

  if (transaction_context_.IsBulkLoad()) {
    FinishBulkLoadProfile();
  }
  transaction_context_.Rollback();

  // Clean up all temporary schemas created through embedded execution.
  // This is safe to do globally because LOAD AS is only supported in
  // READ_WRITE mode, and ConnectionManager enforces that at most ONE
  // read-write connection exists at a time. Therefore, all temporary
  // labels in the schema must belong to this connection.
  execution_slot_->ClearTemporarySchema();
  execution_slot_.reset();
  is_closed_.store(true, std::memory_order_release);

  auto on_close = std::move(on_close_);
  if (on_close) {
    on_close(this);
  }
}

Status Connection::BeginTransaction(TransactionMode mode) {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (transaction_context_.HasOwner()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "A transaction or bulk-load session is already active.");
  }

  try {
    switch (mode) {
    case TransactionMode::kReadOnly:
      transaction_context_.Begin(
          execution_slot_->BeginSnapshotReadTransaction());
      return Status::OK();
    case TransactionMode::kReadWrite: {
      auto transaction = execution_slot_->BeginCurrentCowWriteTransaction();
      if (!transaction) {
        return transaction.error();
      }
      transaction_context_.Begin(std::move(transaction).value());
      return Status::OK();
    }
    }
  } catch (const std::exception& e) {
    return Status::InternalError(
        std::string("Failed to begin explicit transaction: ") + e.what());
  } catch (...) {
    return Status::InternalError("Failed to begin explicit transaction");
  }
  return Status(StatusCode::ERR_INVALID_ARGUMENT,
                "Unsupported explicit transaction mode.");
}

Status Connection::BeginBulkLoad() {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (transaction_context_.HasOwner()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "A transaction or bulk-load session is already active.");
  }

  bulk_load_started_at_ = std::chrono::steady_clock::now();
  try {
    auto transaction = execution_slot_->BeginCurrentCowWriteTransaction();
    if (!transaction) {
      bulk_load_started_at_.reset();
      if (transaction.error().error_code() == StatusCode::ERR_NOT_SUPPORTED ||
          transaction.error().error_code() ==
              StatusCode::ERR_INVALID_ARGUMENT) {
        return Status(transaction.error().error_code(),
                      "Bulk load is supported only in embedded read-write "
                      "mode.");
      }
      return transaction.error();
    }
    transaction_context_.BeginBulkLoad(std::move(transaction).value());
    return Status::OK();
  } catch (const std::exception& e) {
    bulk_load_started_at_.reset();
    return Status::InternalError(std::string("Failed to begin bulk load: ") +
                                 e.what());
  } catch (...) {
    bulk_load_started_at_.reset();
    return Status::InternalError("Failed to begin bulk load");
  }
}

result<QueryResult> Connection::ExecuteBulkLoadQuery(
    const std::string& query_string, const std::string& access_mode,
    const rapidjson::Value& parameters) {
  if (IsClosed()) {
    RETURN_ERROR(
        Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed."));
  }
  if (!transaction_context_.IsBulkLoad()) {
    RETURN_ERROR(Status(StatusCode::ERR_TX_STATE_CONFLICT,
                        "No bulk-load session is active."));
  }
  if (transaction_context_.IsRollbackOnly()) {
    RETURN_ERROR(Status(StatusCode::ERR_TX_STATE_CONFLICT,
                        "Bulk-load session failed; RollbackBulkLoad() is "
                        "required."));
  }

  AccessMode requested_mode;
  try {
    requested_mode = ParseAccessMode(access_mode);
  } catch (...) {
    transaction_context_.AbortAndMarkRollbackOnly();
    throw;
  }
  return execution_slot_->ExecuteBulkLoadQuery(query_string, requested_mode,
                                               parameters, /*num_threads=*/0,
                                               transaction_context_);
}

Status Connection::CommitBulkLoad() {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (!transaction_context_.IsBulkLoad()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "No bulk-load session is active.");
  }
  if (transaction_context_.IsRollbackOnly()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "Bulk-load session failed; RollbackBulkLoad() is required.");
  }
  auto status = execution_slot_->CommitBulkLoad(transaction_context_);
  if (status.ok()) {
    FinishBulkLoadProfile();
  }
  return status;
}

Status Connection::RollbackBulkLoad() {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (!transaction_context_.IsBulkLoad()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "No bulk-load session is active.");
  }
  transaction_context_.Rollback();
  FinishBulkLoadProfile();
  return Status::OK();
}

Status Connection::Commit() {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (transaction_context_.IsBulkLoad()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  transaction_context_.IsRollbackOnly()
                      ? "Bulk-load session failed; RollbackBulkLoad() is "
                        "required."
                      : "Bulk-load sessions must use CommitBulkLoad().");
  }
  if (transaction_context_.IsRollbackOnly()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "Transaction is rollback-only; Rollback() is required.");
  }
  if (!transaction_context_.IsActive()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "No explicit transaction is active.");
  }
  return transaction_context_.Commit();
}

Status Connection::Rollback() {
  if (IsClosed()) {
    return Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed.");
  }
  if (!transaction_context_.HasOwner()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "No explicit transaction is active.");
  }
  if (transaction_context_.IsBulkLoad()) {
    return Status(StatusCode::ERR_TX_STATE_CONFLICT,
                  "Bulk-load sessions must use RollbackBulkLoad().");
  }
  transaction_context_.Rollback();
  return Status::OK();
}

result<QueryResult> Connection::Query(const std::string& query_string,
                                      const std::string& access_mode,
                                      const rapidjson::Value& parameters) {
  VLOG(1) << "Query: " << query_string;
  if (IsClosed()) {
    LOG(ERROR) << "Connection is closed, cannot execute query.";
    RETURN_ERROR(
        Status(StatusCode::ERR_CONNECTION_CLOSED, "Connection is closed."));
  }
  if (transaction_context_.IsRollbackOnly()) {
    if (transaction_context_.IsBulkLoad()) {
      RETURN_ERROR(Status(StatusCode::ERR_TX_STATE_CONFLICT,
                          "Bulk-load session failed; RollbackBulkLoad() is "
                          "required."));
    }
    RETURN_ERROR(
        Status(StatusCode::ERR_TX_STATE_CONFLICT,
               "Transaction is rollback-only; Rollback() is required."));
  }
  if (transaction_context_.IsActive()) {
    if (transaction_context_.IsBulkLoad()) {
      RETURN_ERROR(
          Status(StatusCode::ERR_TX_STATE_CONFLICT,
                 "Bulk-load sessions must use ExecuteBulkLoadQuery()."));
    }
    AccessMode requested_mode;
    try {
      requested_mode = ParseAccessMode(access_mode);
    } catch (...) {
      transaction_context_.AbortAndMarkRollbackOnly();
      throw;
    }
    return execution_slot_->ExecuteQueryInTransaction(
        query_string, requested_mode, parameters, /*num_threads=*/0,
        transaction_context_);
  }
  return execution_slot_->ExecuteQuery(query_string, access_mode, parameters);
}

void Connection::FinishBulkLoadProfile() noexcept {
  if (!bulk_load_started_at_) {
    return;
  }
  const auto elapsed = std::chrono::duration<double>(
      std::chrono::steady_clock::now() - *bulk_load_started_at_);
  try {
    profiling::LoadProfiler::Instance().Record("bulk_load.session.total",
                                               elapsed.count());
  } catch (...) {
    // Profiling must not interfere with commit, rollback, or connection close.
  }
  bulk_load_started_at_.reset();
}

}  // namespace neug

File: tools/python_bind/neug/connection.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# Copyright 2020 Alibaba Group Holding Limited. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

"""The Neug connection module."""

from __future__ import annotations

import logging
from typing import Any
from typing import Dict
from typing import Optional

try:
    from neug_py_bind import PyConnection
except ImportError as e:
    import os

    if os.environ.get("BUILD_DOC", "OFF") == "OFF":
        # re-raise the import error if building documentation
        raise e

from neug.bulk_load_session import BulkLoadSession
from neug.proto.error_pb2 import ERR_CONNECTION_CLOSED
from neug.proto.error_pb2 import OK
from neug.proto.error_pb2 import Code

# This is the C++ binding for the Python interface, which provides the actual connection to the database.
from neug.query_result import QueryResult
from neug.utils import is_access_mode_valid
from neug.utils import valid_access_modes

logger = logging.getLogger(__name__)


class Connection(object):
    """
    Connection represents a logical connection to a database. User should use this class to interact
    with the database, such as executing queries and managing transactions.
    The connection is created by the `Database.connect` method, and should be closed by calling the `close` method
    when it is no longer needed. If the database is closed, all the connections to the database will be closed automatically.
    """

    def __init__(self, py_connection):  # py_connection: PyConnection
        """
        Initialize a Connection object.
        Parameters
        ----------
        py_connection : PyConnection
            The underlying c++ connection object that provides the actual database connection.
        """
        self._py_connection = py_connection
        self._is_open = True

    def __enter__(self):
        """
        Enter the connection context. This method is called when the connection is used in a `with` statement.
        It returns self, so that the connection can be used in the `with` statement.
        """
        return self

    def __exit__(self, exc_type, exc_value, traceback):
        """
        Exit the connection context. This method is called when the `with` statement is exited.
        It closes the connection if it is still open.
        """
        if self._is_open:
            self.close()

    def __del__(self):
        if self._is_open:
            self.close()

    @property
    def is_open(self) -> bool:
        """
        Check if the connection is open.
        Returns
        -------
        bool
            True if the connection is open, False otherwise.
        """
        return self._is_open

    def close(self):
        """
        Close the connection. An active explicit transaction or bulk-load
        session is rolled back.
        """
        if self._is_open:
            self._py_connection.close()
            self._is_open = False

    @property
    def has_active_transaction(self) -> bool:
        """Whether this connection has an active explicit transaction.

        The property remains true while a failed transaction is rollback-only.
        Call :meth:`rollback` to return the connection to auto-commit mode.
        """
        return self._is_open and self._py_connection.has_active_transaction

    def begin_transaction(self, read_only: bool = False):
        """Begin an explicit embedded AP transaction.

        Parameters
        ----------
        read_only : bool
            Pin one read view and reject writes when true. The default starts a
            read-write transaction with a private COW view.

        Raises
        ------
        RuntimeError
            If the connection is closed or already has an active transaction.
        """
        if not self._is_open:
            raise RuntimeError(
                f"Connection is closed. Please open the connection before beginning a transaction. "
                f"Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.begin_transaction(read_only)

    @property
    def has_active_bulk_load(self) -> bool:
        """Whether this connection owns an unfinished bulk-load session."""
        return self._is_open and self._py_connection.has_active_bulk_load

    def begin_bulk_load(self) -> BulkLoadSession:
        """Begin an exclusive session for persistent ``COPY FROM`` queries."""
        if not self._is_open:
            raise RuntimeError(
                "Connection is closed. Please open the connection before "
                f"beginning a bulk load. Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.begin_bulk_load()
        return BulkLoadSession(self)

    def bulk_load(self) -> BulkLoadSession:
        """Return a context-managed bulk-load session."""
        return self.begin_bulk_load()

    def commit(self):
        """Commit the active explicit transaction.

        A rollback-only transaction must be rolled back instead.
        """
        if not self._is_open:
            raise RuntimeError(
                f"Connection is closed. Please open the connection before committing a transaction. "
                f"Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.commit()

    def rollback(self):
        """Roll back the active explicit transaction and return to auto-commit."""
        if not self._is_open:
            raise RuntimeError(
                f"Connection is closed. Please open the connection before rolling back a transaction. "
                f"Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.rollback()

    def execute(
        self, query: str, access_mode="", parameters: Optional[Dict[str, Any]] = None
    ) -> QueryResult:
        """
        Execute a cypher query on the database. User could specify multiple queries in a single string,
        separated by semicolons. The query will be executed in the order they are specified.
        If any query fails, the whole execution will be rolled back.
        If the query is a DDL query, such as `CREATE TABLE`, `DROP TABLE`, etc., the database will be
        modified accordingly.

        For the details of the query syntax, please refer to the documentation of cypher manual.
        The result of the query will be returned as a `QueryResult` object, which contains the result of
        the query and the metadata of the query.
        The QueryResult object is like an iterator, providing methods to iterate over the results,
        such as `__iter__` and `__next__`.

        If the query is a DDL or DML query, the result will be an empty `QueryResult` object.

        Inside an explicit transaction, a failed query leaves the transaction
        rollback-only. Call :meth:`rollback` before executing another query.

        Some of the cypher queries could change the state of the database, such as `CREATE TABLE`, `INSERT`,
        `UPDATE`, `DELETE`, etc. Other queries, such as `MATCH(n) RETURN n.id`, will not change the state of
        the database, but will return the results of the query.

        If the database is opened in read-only mode, any DDL or DML query will raise an exception.
        If the database is opened in read-write mode, all queries could be executed, and the state of the
        database will be changed accordingly.

        .. code:: python

            >>> from neug import Database
            >>> db = Database("/tmp/test.db", mode="w")
            >>> conn = db.connect()
            >>> res = conn.execute('CREATE TABLE person(id INT64, name STRING);')
            >>> res = conn.execute('CREATE TABLE knows(FROM person TO person, weight DOUBLE);')
            >>> res = conn.execute('COPY person FROM "person.csv"')
            >>> res = conn.execute('COPY knows FROM "knows.csv" (from="person", to="person");')
            >>> res = conn.execute('MATCH(n) RETURN n.id')
            >>> for record in res:
            >>>    print(record)
            >>> res = conn.execute('MATCH(p:person)-[knows]->(q:person) RETURN p.id, q.id LIMIT 10;')
            >>> # submitting query with parameters
            >>> res = conn.execute(
                'MATCH (n:person) WHERE n.id = $id RETURN n.name', access_mode='r', parameters={'id': 12345})


        Parameters
        ----------
        query : str
            The query to execute.
        access_mode : str
            The access mode of the query. It could be `read(r)`, `insert(i)`, `update(u)` (include deletion),
            or `schema(s)` for schema modifications. User should specify the correct access mode for the query
            to ensure the correctness of the database. If the access mode is not specified, it is inferred from
            the query text. Supported access modes are:
            - `read`,`r`,`READ`,`R`: for read-only queries
            - `insert`,`i`,`INSERT`,`I`: for insert-only queries
            - `update`,`u`,`UPDATE`,`U`: for update queries (include deletion)
            - `schema`,`s`,`SCHEMA`,`S`: for schema modification operations
        parameters : dict[str, Any] | None
            The parameters to be used in the query. The parameters should be a dictionary, where the keys are the
            parameter names, and the values are the parameter values. If no parameters are needed, it can be set to None.

        Returns
        -------
        query_result : QueryResult
            The result of the query.
        """
        return self._execute(query, access_mode, parameters, bulk_load=False)

    def _execute_bulk_load(
        self, query: str, access_mode="", parameters: Optional[Dict[str, Any]] = None
    ) -> QueryResult:
        return self._execute(query, access_mode, parameters, bulk_load=True)

    def _commit_bulk_load(self):
        if not self._is_open:
            raise RuntimeError(
                "Connection is closed. Please open the connection before "
                f"committing a bulk load. Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.commit_bulk_load()

    def _rollback_bulk_load(self):
        if not self._is_open:
            raise RuntimeError(
                "Connection is closed. Please open the connection before "
                f"rolling back a bulk load. Error code: {ERR_CONNECTION_CLOSED}"
            )
        self._py_connection.rollback_bulk_load()

    def _execute(
        self,
        query: str,
        access_mode: str,
        parameters: Optional[Dict[str, Any]],
        bulk_load: bool,
    ) -> QueryResult:
        if not self._is_open:
            raise RuntimeError(
                "Connection is closed. Please open the connection before "
                f"executing queries. Error code: {ERR_CONNECTION_CLOSED}"
            )
        if access_mode != "" and access_mode.lower() not in [
            "read",
            "r",
            "insert",
            "i",
            "update",
            "u",
            "schema",
            "s",
        ]:
            raise ValueError(
                f"Invalid access_mode: {access_mode}. Supported access modes are "
                f"{valid_access_modes}."
            )
        execute = (
            self._py_connection.execute_bulk_load
            if bulk_load
            else self._py_connection.execute
        )
        ret = QueryResult(execute(query, access_mode, parameters or {}))
        status_code = ret._result.status_code()
        try:
            msg = ret._result.status_message()
        except UnicodeDecodeError:
            msg = "Failed to decode the error message returned from engine"

        if status_code == OK:
            return ret
        else:
            raise RuntimeError(
                f"Failed to execute query: {query}. "
                f"Error code: {status_code}, Error Message: "
                f"{Code.keys()[Code.values().index(status_code)]}: {msg}"
            )

    def get_schema(self):
        """
        Get the schema of the NeuG database.

        :return: The schema of the NeuG database.
        """
        if not self._is_open:
            raise RuntimeError(
                f"Connection is closed. Please open the connection before getting the schema. "
                f"Error code: {ERR_CONNECTION_CLOSED}"
            )
        return self._py_connection.get_schema()
